src/connections.rs

```diff
@@ -240,38 +191,70 @@ impl IncomingMessages {
                 warn!("Failed to read incoming message on bi-stream: {}", err);
                 None
             }
-            Some(Ok((send, recv))) => Self::read_bytes(recv, max_msg_size)
+            Some(Ok((send, mut recv))) => read_bytes(&mut recv, max_msg_size)
                 .await
-                .map(|bytes| (bytes, send)),
+                .ok()
+                .map(|bytes| (bytes, send, recv)),
         }
     }
+}
 
-    // Read the message's bytes which size is capped
-    async fn read_bytes(recv: quinn::RecvStream, max_msg_size: usize) -> Option<Bytes> {
-        match recv.read_to_end(max_msg_size).await {
-            Ok(wire_bytes) => {
-                trace!("Got new message with {} bytes.", wire_bytes.len());
-                match WireMsg::from_raw(wire_bytes) {
-                    Ok(WireMsg::UserMsg(msg_bytes)) => Some(Bytes::copy_from_slice(&msg_bytes)),
-                    Ok(WireMsg::EndpointEchoReq) | Ok(WireMsg::EndpointEchoResp(_)) => {
-                        // TODO: handle the echo request/response message
-                        warn!("UNIMPLEMENTED: echo message type not supported yet");
-                        None
-                    }
-                    Err(err) => {
-                        warn!("Failed to parse received message bytes: {}", err);
-                        None
-                    }
-                }
-            }
-            Err(err) => {
-                warn!("Failed reading message's bytes: {}", err);
-                None
-            }
+/// Read the message's bytes which size is capped
+pub async fn read_bytes(recv: &mut quinn::RecvStream, _max_msg_size: usize) -> Result<Bytes> {
+    let mut data_len: [u8; 1] = [0; 1];
+    recv.read_exact(&mut data_len).await?;
```

Comment:
Why do we call `read_exact` twice(here with a slice and below with a vec)? Recon it's for the flag?

Reply:
This is to obtain the length of the message. 

Although, @lionel1704, I think this is where we should use the `WireMsg`, that struct is already being used for prefixing the message bytes with information which is handled only by the quic-p2p layer, same as the length we need now, so we should add a `msg_length` field (fixed size) to `WireMsg` and do a first read from the stream of length == size_of(<`WireMsg` struct>). Once you read that, you parse it with `WireMsg::from_raw(data)` and from there you obtain the info about message type and message length. Then if what you have is `WireMsg::UserMsg({ msg_bytes, msh_length })` you can read a second time from the stream with a slice/Vec of size == msg_length.

Reply:
>  do a first read from the stream of length == size_of(<WireMsg struct>)

But, this length isn't fixed. So we wouldn't really know how many bytes to read?

Reply:
You are right, the msg bytes should be removed from it, so the WireStruct should have only the length field.